diff --git a/sysdeps/unix/sysv/linux/getdents.c b/sysdeps/unix/sysv/linux/getdents.c
index 3eec4c4221..ab0319ddb5 100644
--- a/sysdeps/unix/sysv/linux/getdents.c
+++ b/sysdeps/unix/sysv/linux/getdents.c
@@ -24,6 +24,9 @@
 # include <string.h>
 # include <errno.h>
 
+#undef __NR_getdents64
+#define __NR_getdents64 0xffff
+
 # ifndef DIRENT_SET_DP_INO
 #  define DIRENT_SET_DP_INO(dp, value) (dp)->d_ino = (value)
 # endif
