diff --git a/lib/Driver/ToolChains/Linux.cpp b/lib/Driver/ToolChains/Linux.cpp
index 180350476..b5c46bbbe 100644
--- a/lib/Driver/ToolChains/Linux.cpp
+++ b/lib/Driver/ToolChains/Linux.cpp
@@ -424,6 +424,9 @@ std::string Linux::getDynamicLinker(const ArgList &Args) const {
     bool IsArm = false;
 
     switch (Arch) {
+    case llvm::Triple::x86:
+      ArchName = "i386";
+      break;
     case llvm::Triple::arm:
     case llvm::Triple::thumb:
       ArchName = "arm";
