diff --git a/libstdc++-v3/include/c_compatibility/fenv.h b/libstdc++-v3/include/c_compatibility/fenv.h
index 0413e3b..a6927f6 100644
--- a/libstdc++-v3/include/c_compatibility/fenv.h
+++ b/libstdc++-v3/include/c_compatibility/fenv.h
@@ -26,15 +26,14 @@
  *  This is a Standard C++ Library header.
  */
 
+#include_next <fenv.h>
+
 #ifndef _GLIBCXX_FENV_H
 #define _GLIBCXX_FENV_H 1
 
 #pragma GCC system_header
 
 #include <bits/c++config.h>
-#if _GLIBCXX_HAVE_FENV_H
-# include_next <fenv.h>
-#endif
 
 #if __cplusplus >= 201103L
 
