diff --git a/gcc/configure b/gcc/configure
index 5a3ef1318..cf7aed959 100755
--- a/gcc/configure
+++ b/gcc/configure
@@ -12262,15 +12262,7 @@ if test x$host != x$target || test "x$TARGET_SYSTEM_ROOT" != x ||
     SYSTEM_HEADER_DIR='$(CROSS_SYSTEM_HEADER_DIR)'
   fi
 
-  if test "x$with_build_sysroot" != "x"; then
-    target_header_dir="${with_build_sysroot}${native_system_header_dir}"
-  elif test "x$with_sysroot" = x; then
-    target_header_dir="${test_exec_prefix}/${target_noncanonical}/sys-include"
-  elif test "x$with_sysroot" = xyes; then
-    target_header_dir="${test_exec_prefix}/${target_noncanonical}/sys-root${native_system_header_dir}"
-  else
-    target_header_dir="${with_sysroot}${native_system_header_dir}"
-  fi
+  target_header_dir="/usr/${target_noncanonical}${native_system_header_dir}"
 else
   target_header_dir=${native_system_header_dir}
 fi
