diff --git a/clang/lib/Driver/ToolChains/Linux.cpp b/clang/lib/Driver/ToolChains/Linux.cpp
index 6532c899492..64a11ac4f3c 100644
--- a/clang/lib/Driver/ToolChains/Linux.cpp
+++ b/clang/lib/Driver/ToolChains/Linux.cpp
@@ -521,6 +521,9 @@ std::string Linux::getDynamicLinker(const ArgList &Args) const {
     bool IsArm = false;
 
     switch (Arch) {
+    case llvm::Triple::x86:
+      ArchName = "i386";
+      break;
     case llvm::Triple::arm:
     case llvm::Triple::thumb:
       ArchName = "arm";
