diff --git a/sysdeps/unix/sysv/linux/getdents.c b/sysdeps/unix/sysv/linux/getdents.c
index e1e2a8b8..ca301cf9 100644
--- a/sysdeps/unix/sysv/linux/getdents.c
+++ b/sysdeps/unix/sysv/linux/getdents.c
@@ -53,7 +53,13 @@ __getdents (int fd, void *buf0, size_t nbytes)
   if (nbytes <= sizeof (struct dirent))
     kbuf = (void*) kbuftmp;
 
-  retval = INLINE_SYSCALL_CALL (getdents64, fd, kbuf, kbytes);
+  // It is affected by "__USE_FILE_OFFSET64" and "__USE_LARGEFILE64".
+  if (sizeof (outp->u.d_ino) != sizeof (inp->k.d_ino) || sizeof (outp->u.d_off) != sizeof (inp->k.d_off)) {
+    retval = INLINE_SYSCALL_CALL (getdents64_x32, fd, kbuf, kbytes);
+  } else {
+    retval = INLINE_SYSCALL_CALL (getdents64, fd, kbuf, kbytes);
+  }
+
   if (retval == -1)
     return -1;
 
